### PR TITLE
Change .templateengine folder path

### PR DESCRIPTION
**Pull request by YuliiaKovalova**
### Problem
Fixes  [#1650](https://github.com/dotnet/templating/issues/1650)

### Solution
Add reading settings path from DOTNET_CLI_HOME if available. 

### Checks
[*] Added unit tests
[ ] Added #nullable enable to all the modified files [?](https://github.com/dotnet/templating/wiki/Contributing#coding-style) N/A

### CODE REVIEW

File: src/Cli/Microsoft.TemplateEngine.Cli/Commands/CliPathInfo.cs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        private static string GetUserProfileDir(IEnvironment environment) => environment.GetEnvironmentVariable(
+            RuntimeInformation.IsOSPlatform(OSPlatform.Windows)
+                ? "USERPROFILE"
+                : "HOME")
+            ?? throw new NotSupportedException("HOME or USERPROFILE environment variable is not defined, the environment is not supported");
+
+        private static string GetGlobalSettingsDir(string? settingsLocation)
+        {
+            var definedSettingsLocation = string.IsNullOrEmpty(settingsLocation)
+                ? CliFolderPathCalculator.DotnetHomePath
```
Comment (vlada-shubina) [inline src/Cli/Microsoft.TemplateEngine.Cli/Commands/CliPathInfo.cs:42]:
```suggestion
                ? Path.Combine(CliFolderPathCalculator.DotnetHomePath, ".templateengine")
```

File: src/Cli/Microsoft.TemplateEngine.Cli/Commands/CliPathInfo.cs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+                : settingsLocation;
+
+            Reporter.Verbose.WriteLine($"Global Settings Location: {definedSettingsLocation}");
+
+            return Path.Combine(definedSettingsLocation, ".templateengine");
```
Comment (vlada-shubina) [inline src/Cli/Microsoft.TemplateEngine.Cli/Commands/CliPathInfo.cs:47]:
```suggestion
            return definedSettingsLocation;
```

File: src/Cli/Microsoft.TemplateEngine.Cli/Commands/CliPathInfo.cs (hunk 1 above)
Comment (vlada-shubina) [reply src/Cli/Microsoft.TemplateEngine.Cli/Commands/CliPathInfo.cs:42]: We don't need to append a directory to `settingsLocation` specified via `debug:custom-settings` arg.

File: src/Cli/Microsoft.TemplateEngine.Cli/Commands/CliPathInfo.cs [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+            var definedSettingsLocation = string.IsNullOrEmpty(settingsLocation)
+                ? Path.Combine(CliFolderPathCalculator.DotnetHomePath, ".templateengine")
+                : settingsLocation;
+
+            Reporter.Verbose.WriteLine($"Global Settings Location: {definedSettingsLocation}");
+
+            return definedSettingsLocation;
+        }
+
+        private static string GetDefaultHostSettingsDir(ITemplateEngineHost host, string? userDir = null, string? globalDir = null)
```
Comment (vlada-shubina) [inline src/Cli/Microsoft.TemplateEngine.Cli/Commands/CliPathInfo.cs:50]: This code may be simplified as `globalDir` is never `null` and userDir is always null.

File: src/Cli/Microsoft.TemplateEngine.Cli/Commands/CliPathInfo.cs [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        {
+            ValidatePathArguments(host, userDir, globalDir);
+
+            return Path.Combine(globalDir ?? GetGlobalSettingsDir(userDir!), host.HostIdentifier);
```
Comment (vlada-shubina) [inline src/Cli/Microsoft.TemplateEngine.Cli/Commands/CliPathInfo.cs:54]:
```suggestion
            return Path.Combine(GlobalSettingsDir, host.HostIdentifier);
```
Not sure if we need additional methods, it looks like doing all the logic constructor will be more readable.

File: src/Tests/dotnet-new.Tests/DotnetNewDebugOptionsTests.cs [hunk 5]
```diff
@@ -121,10 +121,9 @@ public void DoesCreateCacheInDifferentLocationWhenCustomHiveIsUsed()
                .Execute()
                .Should().Pass().And.NotHaveStdErr();
 
-            string[] createdCacheEntries = Directory.GetFileSystemEntries(home);
```
Comment (vlada-shubina) [inline src/Tests/dotnet-new.Tests/DotnetNewDebugOptionsTests.cs:124]: is this change needed?

File: src/Tests/dotnet-new.Tests/DotnetNewDebugOptionsTests.cs [hunk 6]
```diff
@@ -121,10 +121,9 @@ public void DoesCreateCacheInDifferentLocationWhenCustomHiveIsUsed()
...
+            string[] createdCacheEntries = Directory.GetFileSystemEntries(Path.Combine(home));
 
             Assert.Equal(2, createdCacheEntries.Length);
-            Assert.Contains(Path.Combine(home, "packages"), createdCacheEntries);
```
Comment (vlada-shubina) [inline src/Tests/dotnet-new.Tests/DotnetNewDebugOptionsTests.cs:127]: i think it should stay or be changed if needed.

File: src/Tests/dotnet-new.Tests/DotnetNewUninstallTests.cs [hunk 7]
```diff
@@ -198,13 +198,13 @@ public void CanUninstall_NuGet(string commandName)
...
                 .And.HaveStdOutContaining($"Currently installed items:{Environment.NewLine}(No Items)");
 
-            Assert.False(File.Exists(Path.Combine(home, "packages", "Microsoft.DotNet.Web.ProjectTemplates.5.0.5.0.0.nupkg")));
+            Assert.False(File.Exists(Path.Combine(home, ".templateengine", "packages", "Microsoft.DotNet.Web.ProjectTemplates.5.0.5.0.0.nupkg")));
         }
 
         [Fact]
         public void CanUninstallSeveralSources_LegacySyntax()
         {
-            string home = CreateTemporaryFolder(folderName: "Home");
```
Comment (vlada-shubina) [inline src/Tests/dotnet-new.Tests/DotnetNewUninstallTests.cs:207]: should work as is too

File: src/Cli/Microsoft.TemplateEngine.Cli/Commands/CliPathInfo.cs [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+        }
+
+        private static string GetDefaultHostVersionSettingsDir(ITemplateEngineHost host, string? userDir = null, string? globalDir = null)
```
Comment (vlada-shubina) [inline src/Cli/Microsoft.TemplateEngine.Cli/Commands/CliPathInfo.cs:57]: same here: `globalDir` is never null, `userDir` is always null

File: src/Tests/dotnet-new.Tests/DotnetNewUninstallTests.cs [hunk 9]
```diff
@@ -198,7 +198,7 @@ public void CanUninstall_NuGet(string commandName)
                 .NotHaveStdErr()
                 .And.HaveStdOutContaining($"Currently installed items:{Environment.NewLine}(No Items)");
 
-            Assert.False(File.Exists(Path.Combine(home, "packages", "Microsoft.DotNet.Web.ProjectTemplates.5.0.5.0.0.nupkg")));
+            Assert.False(File.Exists(Path.Combine(home, ".templateengine", "packages", "Microsoft.DotNet.Web.ProjectTemplates.5.0.5.0.0.nupkg")));
```
Comment (vlada-shubina) [inline src/Tests/dotnet-new.Tests/DotnetNewUninstallTests.cs:201]: `Assert.False(File.Exists(Path.Combine(home, "packages", "Microsoft.DotNet.Web.ProjectTemplates.5.0.5.0.0.nupkg")));` is correct. 

Since we are using custom settings location, `".templateengine"` won't be appended

File: src/Cli/Microsoft.TemplateEngine.Cli/Commands/CliPathInfo.cs [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+    {
+        public CliPathInfo(
+            ITemplateEngineHost host,
+            IEnvironment environment,
+            string? settingsLocation)
+        {
+            UserProfileDir = GetUserProfileDir(environment);
+            GlobalSettingsDir = GetGlobalSettingsDir(settingsLocation);
+
+            ValidatePathArguments(host, GlobalSettingsDir);
```
Comment (vlada-shubina) [inline src/Cli/Microsoft.TemplateEngine.Cli/Commands/CliPathInfo.cs:22]: `GlobalSettingsDir` cannot be `null`, so second part of validation is not needed.
It is preferable that validation happens as first operation in constructor, to fail immediately in case of error.

File: src/Cli/Microsoft.TemplateEngine.Cli/Commands/CliPathInfo.cs [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+
+        private static string GetUserProfileDir(IEnvironment environment) => environment.GetEnvironmentVariable(
+            RuntimeInformation.IsOSPlatform(OSPlatform.Windows)
+                ? "USERPROFILE"
+                : "HOME")
+            ?? throw new NotSupportedException("HOME or USERPROFILE environment variable is not defined, the environment is not supported");
+
+        private static string GetGlobalSettingsDir(string? settingsLocation)
+        {
+            var definedSettingsLocation = string.IsNullOrEmpty(settingsLocation)
```
Comment (vlada-shubina) [inline src/Cli/Microsoft.TemplateEngine.Cli/Commands/CliPathInfo.cs:48]:
```suggestion
            var definedSettingsLocation = string.IsNullOrWhitespace(settingsLocation)
```

path cannot be a `whitespace`

File: src/Cli/Microsoft.TemplateEngine.Cli/Commands/CliPathInfo.cs [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+            if (string.IsNullOrWhiteSpace(host.HostIdentifier))
+            {
+                throw new ArgumentException($"{nameof(host.HostIdentifier)} of {nameof(host)} cannot be null or whitespace.", nameof(host));
+            }
+            if (string.IsNullOrWhiteSpace(globalDir))
+            {
+                throw new ArgumentException($"{nameof(globalDir)} cannot be null or whitespace.");
+            }
+        }
+    }
```
Comment (vlada-shubina) [inline src/Cli/Microsoft.TemplateEngine.Cli/Commands/CliPathInfo.cs:68]: nit: validation of global directory is not needed.
the host validation may be moved to constructor for code readability, given that version validation happens in constructor. Or move validation of the version here.

File: src/Cli/Microsoft.TemplateEngine.Cli/Commands/CliPathInfo.cs [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+            UserProfileDir = GetUserProfileDir(environment);
+            GlobalSettingsDir = GetGlobalSettingsDir(settingsLocation);
+
+            ValidatePathArguments(host, GlobalSettingsDir);
+            HostSettingsDir = Path.Combine(GlobalSettingsDir, host.HostIdentifier);
+
+            if (string.IsNullOrWhiteSpace(host.Version))
+            {
+                throw new ArgumentException($"{nameof(host.Version)} of {nameof(host)} cannot be null or whitespace.", nameof(host));
+            }
```
Comment (vlada-shubina) [inline src/Cli/Microsoft.TemplateEngine.Cli/Commands/CliPathInfo.cs:28]: It is preferable that validation happens as first operation in constructor, to fail immediately in case of error.

File: src/Cli/Microsoft.TemplateEngine.Cli/Commands/CliPathInfo.cs (hunk 3 above)
Comment (YuliiaKovalova) [reply src/Cli/Microsoft.TemplateEngine.Cli/Commands/CliPathInfo.cs]: Done, please take a look

**Review by vlada-shubina — APPROVED** [review]
Looks good to go!
